src/core_plugins/kibana/public/dashboard/dashboard_state_manager.js

```diff
@@ -138,20 +138,23 @@ export class DashboardStateManager {
   }
 
   _handleStoreChanges() {
-    if (this._areStoreAndAppStatePanelsEqual()) {
-      return;
+    let dirty = false;
+    if (!this._areStoreAndAppStatePanelsEqual()) {
+      const state = store.getState();
+      this.appState.panels = [];
+      _.map(getPanels(state), panel => {
+        this.appState.panels.push(panel);
+      });
+      dirty = true;
     }
 
-    const state = store.getState();
-    // The only state that the store deals with that appState cares about is the panels array. Every other state change
-    // (that appState cares about) is initiated from appState (e.g. view mode).
-    this.appState.panels = [];
-    _.map(getPanels(state), panel => {
-      this.appState.panels.push(panel);
-    });
+    const fullScreen = getFullScreenMode(store.getState());
```

Comment:
FWIW, I get a _little_ lost with the number of ways we're accessing and tracking state. It seems like in some places we use internal  helpers, e.g. line 142 calls an internal helper `this._areStoreAndAppStatePanelsEqual()`. And then in some places like here and line 143, we pull a state object out of the store and pass it to an accessor. If we could simplify this to a consistent pattern I think it would be easier for me to maintain this file if I came into it cold. I know this is beyond this PR's scope, so just some food for thought. :)

Reply:
Oh yes, I can empathize. This file is fragile and confusing.  I def want to clean it up, there are just a lot of moving parts in this area of the code right now, so it's a matter of taking many small steps.  I want to clean up the embeddable stuff (this merged PR has a diagram of where I'm heading: https://github.com/elastic/kibana/pull/14863), and what state gets stored in the url (https://github.com/elastic/kibana/issues/14455), and removing uiState (https://github.com/elastic/kibana/pull/14949).  Then I want to see if I can use react router to connect redux to the url changes rather than doing it manually.  Then, do we need AppState at all?  Or if I can reactify other parts of the top nav panels, then I can move more state into redux and less in AppState/scope, so migrating the top nav to k7 style also affects this.

This PR is also a small part of it. Moving something off scope and into react/redux.

So yea, def on my radar. Just need to take it a bit slow as I navigate cleaning up all these moving parts.

But as for simpler ways to clean this up.... the panel comparison is more complicated bc one is an object, one an array, but I could still probably do something to make it look more similar to the other comparisons.

Reply:
Thanks for the great explanation + links! This is very exciting.